### ISSUE TITLE
[KED-1272] Video and image datasets being read and written to/from memory

### ISSUE THREAD

**Issue by dasturge** [quantumblacklabs/kedro#125]
## Description

I am working with video and image datasets.  As such, I'm either working with images or with sequences of images as raw data.

In looking through the documentation and the current available Dataset objects, the `_load` and `_save` functions seem to want to read and write entire datasets to/from memory.  For images, and especially for videos, this is not feasible.  In large NLP or audio contexts it also may not work, though they may be more straightforward to partition.

One option of course, is to handle data paths instead of loading the raw images.  However, this will break the principle of "pure functions", as it will incur a number of filesystem changes that the user must keep careful track of and are not handled by kedro's api.

In summary, it's unclear how to move forward with very high dimensional data.  I'm looking for general proposals of how to move forward, maybe some input from an experienced kedro user or proposals for framework additions.

## Possible Implementations

1. A basic recipe or notebook example for implementing such a dataset/dag
2. A graph computation solution which involves batching datasets, e.g. some iterload function for datasets, and dag expansion over a dynamically sized iterable, then merging of saved results.  A sort of map-reduce on the dataset.
3. A safe io class, create a filetype class that needs an "input", an "output", reads and saves data.

**Comment by 921kiyo** [comment]
Thank you for opening an issue! This is not perfect solution (and slightly break the "pure function" as you mentioned). You could open a file in `_load` and handle the partition of the file object inside the node function. 

Another solution I could think of is to [create a custom dataset](https://kedro.readthedocs.io/en/latest/03_tutorial/03_set_up_data.html#creating-custom-datasets) using libraries like Dask (example https://github.com/quantumblacklabs/kedro/pull/97) to enable load in chunks.

**Comment by yetudada** [comment]
Hi @dasturge, I've just put this one on our internal tracker so we can have a look at how to handle this. But it would be great to find out, was `dask-image` any help?

**Comment by dasturge** [comment]
Actually, in my case, I eventually realized that what I really needed was a `torch` Dataset loader.

```
class CSVTorchLoader(AbstractDataSet):

    def __init__(self, image_folder: str, csv_file: str,
        load_args: Dict[str, Any]) -> None:
        super().__init__()
        self.image_folder = image_folder
        self.csv_file = csv_file
        _load_args = get_default_kwargs(MyTorchDataset)
        _load_args.update(load_args)
        self._load_args = _load_args

    def _describe(self) -> Dict[str, Any]:
        return {
            'image_folder': self.image_folder,
            'csv_file': self.csv_file,
            'load_args': self._load_args,
        }

    def _load(self) -> MyTorchDataset:
        return MyTorchDataset(
                       self.image_folder, self.csv_file,
                       **self._load_args)

    def _save(self, data) -> None:
        raise AttributeError("can't save Torch Datasets")

def get_default_kwargs(callable):
    signature = inspect.signature(callable)
    return {
        k: v.default
        for k, v in signature.parameters.items()
        if v.default is not inspect.Parameter.empty
    }
```

So...  I dropped the `dask-image` route shortly after writing a function for it.

`dask-image` will be pretty easy to implement.  I didn't test around with it a whole lot but in principle it works for image processing purposes.  

It's a little clunky to use `dask-image` along with `pillow` though, and for that purpose, `PIL.Image.open` is actually a lazy operation.  With a simple use of temp space, and a simple wrapper class for saving PIL images and reopening after applying any sequence of operations, it should actually be pretty simple to open a collection of `PIL` images as well.

**Comment by yetudada** [comment]
Hi @dasturge! We actually released an [`ImageDataSet`](https://kedro.readthedocs.io/en/latest/kedro.extras.datasets.pillow.ImageDataSet.html) that uses `pillow` recently. You may want to check that out. I'll close this issue for now, let us know if you need anything else.
